### PR TITLE
Merge latex targets

### PR DESCRIPTION
**Pull request by mathijshenquet**
Merged the latex targets together. latex2unicode is now a many to many mapping, combining the codes from: latex, varlatex (variant latex code), mathlatex (math variant latex code) and ams (latex code from the popular amsmath package). 

Also started to make a variant of unicode.xml removing entries not relevant for the latex2unicode mapping (for example \mathbb{R} as its not a single latex entry).

### CODE REVIEW

**Comment by digitalheir** [comment]
h hey, thanks good stuff

But can u please spread the changes over multiple pull requests, each solving a single issue. Some of these changes I agree with, some I'm unsure about.

File: java/src/main/java/com/github/digitalheir/unicode/Character.java [hunk 1]
```diff
@@ -181,16 +155,6 @@ public String getType() {
   }
 
   /**
-   * Sets the value of the type property.
```
Comment (digitalheir) [inline java/src/main/java/com/github/digitalheir/unicode/Character.java]: Why did you remove all these setters? iirc JAXB unmarshalls XML into classes it expects to be JavaBeans. Maybe it's OK because the fields are public? Anyway, this is generated code so be cautious.

File: java/src/main/java/com/github/digitalheir/unicode/Character.java [hunk 2]
```diff
@@ -280,14 +218,19 @@ public String getMathlatex() {
...
 
-  /**
-   * Sets the value of the mathlatex property.
-   *
-   * @param value allowed object is
-   *              {@link String }
-   */
-  public void setMathlatex(String value) {
-    this.mathlatex = value;
+  public String[] getLatexCodes() {
```
Comment (digitalheir) [inline java/src/main/java/com/github/digitalheir/unicode/Character.java]: So do we lose information about which package / mode to use with the control sequence?  Is relevant information.

File: java/src/main/resources/com.github.digitalheir.unicode/charlist.dtd [hunk 3]
```diff
@@ -1,73 +0,0 @@
-<!ELEMENT charlist (entitygroups?,mathvariants?,character*)>
```
Comment (digitalheir) [inline java/src/main/resources/com.github.digitalheir.unicode/charlist.dtd]: Why delete this information?

**Comment by digitalheir** [comment]
> Also started to make a variant of unicode.xml removing entries not relavant for the latex2unicode mapping (for example \mathbb{R} as its not a single latex entry).

Speaking on this, I don't understand what you want to achieve. It's a trivial operation to filter what you want out of the source data?

**Comment by digitalheir** [comment]
(no bully intended)

File: java/src/main/java/com/github/digitalheir/unicode/Character.java (hunk 1 above)
Comment (mathijshenquet) [inline java/src/main/java/com/github/digitalheir/unicode/Character.java]: No it uses the fields directly

File: java/src/main/resources/com.github.digitalheir.unicode/charlist.dtd (hunk 3 above)
Comment (mathijshenquet) [inline java/src/main/resources/com.github.digitalheir.unicode/charlist.dtd]: It was double in the repo

Comment (mathijshenquet) [inline java/src/main/resources/com.github.digitalheir.unicode/charlist.dtd]: Oh was one the w3c original? That wasn't really clear to me

File: java/src/main/java/com/github/digitalheir/unicode/Character.java (hunk 2 above)
Comment (mathijshenquet) [inline java/src/main/java/com/github/digitalheir/unicode/Character.java]: Well lets look at this: 

* Most occurrences of <mathlatex> are ways of writing the same character in mathmode, but characters that are _only_ used in mathmode dont have a <mathlatex> tag. Instead you should probably look at a combination of the <character mode> property and the <mathlatex> or <latex> tag to determine the modes.

* The <varlatex> tags sometimes provide a good alternative, like the alternative \bot (bottom element) for \perp (perpendicular). However most of the time the alternative is just wrong or inaccurate. What happens a lot is that there is a variant of \geq using two lines called \geqq both of these have different unicode points and a present in the file. But the \geq character has \geqq as alternative code causing the \geq entry in the latex2unicode to have a wrong entry.

* The <ams> codes usually are the same as the <latex> codes. AMS has extra symbols like /iiint (double integral) not present in plain latex so its a clear extension and therefore mergeable.

File: java/src/main/java/com/github/digitalheir/ConvertToJson.java [hunk 4]
```diff
@@ -280,8 +277,59 @@ private static void mapAndWriteGsonFiles(List<Character> characters, String stri
                                         ? s.getValue().iterator().next()
                                         : s.getValue()
                         )
+                );
+    }
+
+    public static <T, U, A, R>
+    Collector<T, ?, R> flatMapping(Function<? super T, U[]> mapper,
```
Comment (mathijshenquet) [inline java/src/main/java/com/github/digitalheir/ConvertToJson.java]: Couldnt find it

File: java/src/main/java/com/github/digitalheir/unicode/Character.java (hunk 1 above)
Comment (digitalheir) [inline java/src/main/java/com/github/digitalheir/unicode/Character.java]: kthx

File: java/src/main/java/com/github/digitalheir/unicode/Character.java (hunk 2 above)
Comment (digitalheir) [inline java/src/main/java/com/github/digitalheir/unicode/Character.java]: 1. kthnx, but then why flatten all in an array
2. i guess varlatex doesn't need to be the exact same unicode character, but somehow related. Ironically symbols like `\epsilon` and `\varepsilon` do not refer to each other though
3. clear extension? it still seems you are dropping information when a. i find that `\iint` as a result for my query. how would i know that i need the ams package before LaTeX compiler complains?

Concluding, I regard the XML file to be authoritative unless it is very obviously wrong.

File: java/src/main/resources/com.github.digitalheir.unicode/charlist.dtd (hunk 3 above)
Comment (digitalheir) [inline java/src/main/resources/com.github.digitalheir.unicode/charlist.dtd]: I think my mistake for having a duplicate file

File: java/src/main/java/com/github/digitalheir/unicode/Character.java (hunk 2 above)
Comment (mathijshenquet) [inline java/src/main/java/com/github/digitalheir/unicode/Character.java]: 2. But funnily enough GREEK SMALL LETTER EPSILON has latex code \epsilon and ams code \varepsilon, i think this is one of the few places they disagree.

Comment (digitalheir) [inline java/src/main/java/com/github/digitalheir/unicode/Character.java]: please be patient

Comment (digitalheir) [inline java/src/main/java/com/github/digitalheir/unicode/Character.java]: 2. is `\varepsilon` only in AMS or in vanilla LaTeX as well? I'm suspecting the table has a lot of strange mappings like this. 

Starting to think: yes, maybe good to have a clear opinion & start correcting these spurious mappings.

BTW I've been using [this test file](https://github.com/digitalheir/latex-to-unicode-converter/blob/master/test/mathy.ts) to see how much the mappings in the W3C unicode.xml file agree with my other project `latex-to-unicode-converter`: I suspect many more.

Comment (digitalheir) [inline java/src/main/java/com/github/digitalheir/unicode/Character.java]: After looking at the occurrence of `varlatex`, I completely agree with you. It's only used 18 times anyway.

Comment (digitalheir) [inline java/src/main/java/com/github/digitalheir/unicode/Character.java]: After some more inspection, indeed field names are strange. e.g., [`<latex>\mathbb{C}</latex>`](https://digitalheir.github.io/mathy-unicode-characters/?q=mathbb{c}) is unexpected on two accounts:

1. `\mathbb` requires ams, so i'd expect it to be listed under `ams`
2. `\mathbb` only works in math mode, so i'd expect it to be listed under `mathlatex`

Kind of surprised by these grave inaccuracies. Agree it's better to purge these unwanted elements to keep our data set pure.

File: java/src/main/java/com/github/digitalheir/ConvertToJson.java (hunk 4 above)
Comment (digitalheir) [inline java/src/main/java/com/github/digitalheir/ConvertToJson.java]: rekt

**Comment by mathijshenquet** [comment]
Hey Maarten, 

In my latest commit I recreated the unicode.latex.xml file this time only making corrections to \<latex\> and \<varlatex\> we discussed above. Here is a [full diff wrt the base unicode.xml](https://gist.github.com/mathijshenquet/19e3c6ca0218da466c3eabc36091562c). 

I also added a mechanism for detecting non canonical (i.e. composite, like `\not\subset` or `\mathbb{R}`) latex codes (adapted from your other package). This allows us to create a latex2unicode mapping containing only pure latex commands (i.e. consisting of a single command like `\subsetneqq`), which is the usecase i'm most interested in. All the while retaining a good unicode2latex mapping where there might be only a composite latex representation. My rational for this is that the latex2unicode mapping should be for single tokens: if you had a composite latex term like \not\subset you would want to parse it first.

As you can [see in latex2unicode.json generated from unicode.latex.xml](https://github.com/mathijshenquet/mathy-unicode-characters/blob/master/json/latex2unicode.json) there are only a few ambiguities remaining where there is no clear preference (I also added pretty printing for the json).

Greetings

File: java/src/main/java/com/github/digitalheir/unicode/Character.java (hunk 2 above)
Comment (mathijshenquet) [inline java/src/main/java/com/github/digitalheir/unicode/Character.java]: What even more weird is letters like MATHEMATICAL BOLD CAPITAL MU having `<latex>M</latex>` while other characters in the MATHEMAICAL BOLD range have `<latex>\mathbf{...}</latex>`. Presumably because latex has no command `\Mu`.

**Comment by digitalheir** [comment]
Good, I'll check it out & prolly merge when I get home.

In a sense `\not\subset` is a canonical form to represent ⊄. True, in LaTeX it's  a compound of `¬` and `⊂` (ideographically), and maybe it can also be written as `\not{\subset}` etc. but in Unicode it's a single glyph. It's not obvious that these two control sequences are needed to generate this particular Unicode character. 

Conversely, note that [there are cases](https://digitalheir.github.io/mathy-unicode-characters/?q=NotEqualTilde) where a single (0-arg) LaTeX command generates a succession of Unicode characters (≂̸).

I see no problem in any this, except it might obscure the fact that they are more accurately described as functions and you can't enumerating exhaustively commands like `\`{}`. 

In any case, I very much like the idea of adding a flag "this is a complex LaTeX expression".

File: .gitignore [hunk 5]
```diff
@@ -6,4 +6,3 @@ java/target/
 ts-compiled/
 *.log
 /ts/
-typescript/package-lock.json
```
Comment (digitalheir) [inline .gitignore]: nothing personnel.... kid

File: java/src/main/java/com/github/digitalheir/unicode/Character.java (hunk 2 above)
Comment (digitalheir) [inline java/src/main/java/com/github/digitalheir/unicode/Character.java]: Um. There's still `\mathbf{M}`. Maybe we can conclude that the W3C sauce is rife with errors... I think we should focus on making [`latex-to-unicode-converter`](https://github.com/digitalheir/latex-to-unicode-converter) as complete as possible, and think of a structured way to dump all supported commands as a big JSON file. This project was orginally meant as a validation tool anyway.

What do u think friend? 

(There's still some commands in this list that `latex-to-unicode-converter`, and also vice versa, for instance `\frac` and `\sqrt`)

https://digitalheir.github.io/latex-to-unicode-converter/

**Comment by digitalheir** [comment]
In the feature pls make your pull requests based on features and not "my own master branch", ideally solving an [issue](https://github.com/digitalheir/mathy-unicode-characters/issues). It's pretty hard to review like this - especially so because I don't know what your intentions are
